### DISCUSSION TITLE
PDF 高级打印方案

### DISCUSSION THREAD

**Discussion by Kenshin** [Kenshin/simpread#3887] [同步助手]
# 描述

使用 `wkhtmltopdf` 来增强导出 PDF 的定制化。

# 下载地址

这个是 PDF 辅助增强插件的的功能，需要安装 [对应的插件](https://github.com/Kenshin/simpread/discussions/2721)

# 演示

[百度网盘](https://pan.baidu.com/s/1a2ONidboUASR92N-Xkllmg?pwd=f3y5#list/path=%2FSimpRead%2Fpdf)

# 功能比较

同步助手基础/高级打印方案的区别：

|        | 同步助手    | 依赖               | 优势                         | 劣势            |
|--------|---------|------------------|----------------------------|---------------|
| 基础打印功能 | 1.0.1 版 | 无需多余依赖，仅使用同步助手   | 无需 wkhtmltopdf 依赖          | 阅读模式、稍后读中均可调用 |
| 高级打印功能 | 1.0.2 版 | 需要安装 wkhtmltopdf | 无图片无法显示的问题，且可定制主题、样式等一切内容。 | （暂时）仅能阅读模式下使用 |

# wkhtmltopdf

[wkhtmltopdf](https://wkhtmltopdf.org/) 是一个 HTML → PDF 的第三方库。

## 下载地址

https://wkhtmltopdf.org/downloads.html 请根据自己的环境找到相应安装包。

## 安装

安装过程从略，**但如果你是 Windows 11 的话，建议不要选择默认位置，请选择其它盘。**

## 验证

进入命令行，并执行 `wkhtmltopdf -V` 如果有类似输出的话，即说明安装成功。

![image](https://user-images.githubusercontent.com/81074/172299003-bcfaf031-b2bf-482e-9a9a-42e204c7a3fc.png)

**注意：**

1. 如果没有输出的话，请指明 `wkhtmltopdf` 的安装目录，**尤其是 Windows 用户**。

2. Windows 用户建议安装在非系统盘目录 `Program Files` 下，同时在请插件中手动指定路径。

# 使用

当选择了 **同步助手高级打印方案** 时生效。

![image](https://user-images.githubusercontent.com/81074/172297297-23633dcf-1495-4b4c-bef8-2b3dd5ece036.png)

除此之外， **同步助手高级打印方案** 还提供了众多的定制化功能：

- [主题、字体样式、字体大小](https://github.com/Kenshin/simpread/discussions/3887#discussioncomment-2735661)

- [自定义字体](https://github.com/Kenshin/simpread/discussions/3887#discussioncomment-2735667)

- [自定义样式](https://github.com/Kenshin/simpread/discussions/3887#discussioncomment-2735664)

- [更细致的打印，如页边距、打印尺寸](https://github.com/Kenshin/simpread/discussions/3887#discussioncomment-2750058)

- 自带目录

- 完美解决因图片防盗链无法成功打印的问题。

**Comment by Kenshin** [comment]
# 默认主题

高级打印功能会根据阅读模式设置的主题进行打印，也就说导出的 PDF 会跟你的阅读模式保持一致性，但也可以自主选择其它主题。

![image](https://user-images.githubusercontent.com/81074/172300967-f8ac3a3a-e5c5-4898-a4f0-a1f489e8f6a3.png)

## 一个例子

![image](https://user-images.githubusercontent.com/81074/172302007-47e3734c-0ae0-47d9-8ec6-c9a95d071cfb.png)

**Reply by helloyan** [reply]
同步助手高级打印方案  这个设置再哪里，翻了一圈没找到

**Reply by Kenshin** [reply]
@helloyan 

截图对应的插件在 https://github.com/Kenshin/simpread/discussions/2721 另外，前面的描述中也说明了这个链接。
如果你安装后没有找到的话，试试 https://github.com/Kenshin/simpread/discussions/2342 提到的解决方案。

**Reply by helloyan** [reply]
已找到，在”动作”页面最下面

**Comment by Kenshin** [comment]
# 自定义样式

不止如此，也可以在主题的基础上微调各种参数。

![image](https://user-images.githubusercontent.com/81074/172302282-ea47958d-ee94-4330-8add-ad1ac6334fbc.png)

## 图片大小

大部分网页都可以通过上图所示的设置修改图片大小

<img width="382" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/f17b7d09-1144-48a6-9094-ee17bb1d804b">

但并不是每个网页都可以正常使用此功能，当无法图片大小修改无效时，可以使用自定义样式解决，将下面的代码

```
sr-rd-content img {
  max-width: 50%!important;
}
```

（**注意：`50%` 可以修改为任意尺寸**）加入到下面所示的地方

<img width="588" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/4126c3ef-f0cb-4a63-846d-213cf9aab613">

保存后再打印。

**Reply by DHS1001** [reply]
在这里设定图片宽度后（例如50），在导出的pdf好像不生效。等迟些再试一下其他网站。

**Reply by DHS1001** [reply]
> 在这里设定图片宽度后（例如50），在导出的pdf好像不生效。等迟些再试一下其他网站。

好像图片宽度在这里设置无效呀，但是在“全局样式”设置就可以哦

**Reply by Kenshin** [reply]
@DHS1001 

嗨，这个 Bug Fixed 掉了，细节请看这里 https://github.com/Kenshin/simpread/issues/5412#issuecomment-1684814731

**Comment by Kenshin** [comment]
# 全局字体

![image](https://user-images.githubusercontent.com/81074/172302833-0d1a4f88-5bd8-46ea-9d25-de9f8d7f0a89.png)

可以支持本机任意字体样式或者使用 Google font 设置全局字体。

```
@import url('https://fonts.googleapis.com/css2?family=ZCOOL+XiaoWei&display=swap');

body, body sr-rd-content * {
  font-family: "ZCOOL XiaoWei"!important;
}
```

## 一个例子

下图就是使用上面的全局字体导出的效果。

![image](https://user-images.githubusercontent.com/81074/172302714-29eec4e2-9a1b-4c4b-bd99-d0e6fb2e1bff.png)

**Reply by Bioerjzy** [reply]
请问本地字体怎么使用？

**Reply by Kenshin** [reply]
@Bioerjzy 

只要知道本地字体的 `font-family` 直接写就行。

```
body, body sr-rd-content * {
  font-family: "xxxx"!important;
}
```

将 `xxxx` 换成本机任意 `font-family` 名称。

**Comment by Kenshin** [comment]
# 全局样式

![image](https://user-images.githubusercontent.com/81074/172303070-176e6578-7676-4495-b615-b008a447c63c.png)

可以在这里设置任意全局样式。

**Comment by Kenshin** [comment]
# 标注色

因为 `wkhtmltopdf` 对 CSS 样式支持度的缘故，在此模式下标注色并不能 100% 很好的显示，所以当开启此选项后，无论标注色的样式是什么一律改为统一样式。

![image](https://user-images.githubusercontent.com/81074/172303426-250dd1cb-8de3-4cb6-b346-5be6d9aea480.png)

## 建议

如果对标注没有要求的话，请关闭此选项，关闭后将不会显示标注

**Comment by Kenshin** [comment]
# wkhtmltopdf 定制化

支持边距设定与更多的定制化参数。

![image](https://user-images.githubusercontent.com/81074/172303710-b9260a20-6c2e-444d-ab4d-865bcd552fbc.png)

## 定制化参数

支持 wkhtmltopdf [全部的定制化参数](https://wkhtmltopdf.org/usage/wkhtmltopdf.txt)。

## 一个例子

```
--page-size Letter
```

![image](https://user-images.githubusercontent.com/81074/172304260-375dbdfb-e63c-4d9b-aca3-117279e1cd26.png)

![image](https://user-images.githubusercontent.com/81074/172304095-ccf6a6ae-22f6-4240-9a06-d0c91b5def83.png)

## 在每页加入页脚功能

将 `--footer-right [page]/[topage] --footer-font-size 8 --footer-spacing 3` 粘贴到此处（注意：此方式不能将页边距设置为 `0`）

<img width="581" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/e09397e9-33b7-4cf9-b037-ec49a4115837">

效果

<img width="223" alt="image" src="https://github.com/Kenshin/simpread/assets/81074/3bdaf4d8-c27c-4aa9-9e5a-5aad6d8d971d">

**Reply by Kenshin** [reply]
# 备注

> 页边距除了上述参数影响外，还会受到界面的版面布局的影响。

<img width="1822" alt="image" src="https://user-images.githubusercontent.com/81074/211712268-f7709115-40fb-4d1c-a2fa-2f869d0aa3f3.png">

所以，如果要修改页边距的话，需要先按照上图修改，然后刷新当前页面再打印。

***

除此之外不改变版面布局的话，还有个办法，将下面的代码

```
body {
    margin: 20px 0!important;
}
```

放到下图所示的位置，然后保存。

<img width="607" alt="image" src="https://user-images.githubusercontent.com/81074/211712879-b102fddd-8d8e-4869-b5ef-672c3659e5b0.png">

备注：

- 20px → 上下边距（一般不用改）

- 0 → 左右边距，设置为0或其它 CSS 3 支持的值。

**Comment by Kenshin** [comment]
# 运行目录

指定 `wkhtmltopdf` 的运行目录

1. `Mac` 与 `Linux` 如果默认安装的话，一般来说不需要设置，如设置一般值为 `/usr/local/bin/`
2. `Windows 11` **建议不要安装在默认盘和目录，可能会有权限问题导致调用失败**，所以这是需要指定运行目录，一般是包含 `bin` 的目录结构。

![image](https://user-images.githubusercontent.com/81074/172304346-dfc9da8e-3b3f-476e-a16f-9422d99262c1.png)

**Reply by DHS1001** [reply]
在win7 64位下，使用默认`wkhtmltopdf `的安装目录和放在C盘根目录（均指定运行目录）都提示“导出失败”。
后经各位大佬提示和帮助，把wkhtmltopdf放在非C盘（例如D盘），然后指定运行目录，就成功导出啦。非常感谢！

**Reply by EdisonWendy** [reply]
请问一下Windows指定运行目录也是/usr/local/bin/吗，怎么指定了还是失败
![微信截图_20230927131536](https://github.com/Kenshin/simpread/assets/120798970/3497d9d4-2859-4f77-81db-de01b0fea273)

**Reply by Kenshin** [reply]
@EdisonWendy 

`/usr/local/bin/` 是 Mac 下面的地址.... Windows 在安装时有它自己的地址，或者在命令行（似乎）也能看安装地址

**Reply by EdisonWendy** [reply]
懂了

**Comment by Kenshin** [comment]
# 临时目录

大部分用户都不需要开启此选项，当你使用 Onedrive 作为导出目录时建议开启，详细说明 [请看这里](https://github.com/Kenshin/simpread/issues/3707)。

**Comment by Kenshin** [comment]
RT
